server/src/main/java/org/opensearch/rest/NamedRoute.java

```diff
@@ -9,47 +9,140 @@
 package org.opensearch.rest;
 
 import org.opensearch.OpenSearchException;
+import org.opensearch.transport.TransportService;
+
+import java.util.HashSet;
+import java.util.Set;
 
 /**
  * A named Route
  *
  * @opensearch.internal
  */
 public class NamedRoute extends RestHandler.Route {
+
     private static final String VALID_ACTION_NAME_PATTERN = "^[a-zA-Z0-9:/*_]*$";
     static final int MAX_LENGTH_OF_ACTION_NAME = 250;
 
-    private final String name;
+    private final String uniqueName;
+    private final Set<String> actionNames;
 
-    public boolean isValidRouteName(String routeName) {
-        if (routeName == null || routeName.isBlank() || routeName.length() > MAX_LENGTH_OF_ACTION_NAME) {
-            return false;
+    /**
+     * Builder class for constructing instances of {@link NamedRoute}.
+     */
+    public static class Builder {
+        private RestRequest.Method method;
+        private String path;
+        private String uniqueName;
+        private Set<String> legacyActionNames;
```

Comment:
```suggestion
        private Set<String> legacyActionNames = new HashSet<String>();
```

Then no more null checks needed

Reply:
What if the routes don't register any legacyActionNames, this would lead to creation of empty sets for all of those routes.

Reply:
Yes, that seems fine.  You have concerns?

Reply:
I guess my only concern was with memory but this doesn't seem to cause a big impact. I will make the change.